cuML/src/glm/qn/qn_solvers.h

```diff
@@ -138,8 +138,9 @@ inline OPT_RETCODE min_lbfgs(const LBFGSParam<T> &param,
       return OPT_NUMERIC_ERROR;
     }
     // Save the curent x and gradient
-    xp = x;
-    gradp = grad;
+    xp.copy_async(x, stream);
+    gradp.copy_async(grad, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```

Comment:
are the `copy_async` of `x` to `xp` and `grad` to `gradp` device to device copies? If so I think the `cudaStreamSynchronize` is not really needed as subsequent operations are submitted in the same stream, right?

Reply:
Yes, it's `DeviceToDevice`. 
I think you're right that we can remove the syncs pointed out here and below.
Sync points are in principle quite well defined, i.e. whenever a scalar is computed and transferred to the host. 
I guess I have to get a bit more familiar with the whole stream mechanics